### ISSUE TITLE
[feature] Better management for well known deleted options and settings

### ISSUE THREAD

**Issue by uilianries** [conan-io/conan#11983]
With the introduction of very explicit values management on Conan 2.0, we will need to protect false deletion for options like `fPIC` and compiler attributes for `msvc`. As real example:

```python
def config_options(self):
    if self.settings.os == "Windows":
        del self.options.fPIC

def configure(self):
    if self.options.shared:
        try:
            del self.options.fPIC
        expect Exception:
            pass
    try:
        del self.settings.compiler.cppstd
    except Exception:
        pass
    try:
        del self.settings.compiler.libcxx
    except Exception:
        pass
```

When using the compiler `msvc`, it will discard `cppstd` and `libcxx`, so it's needed to protect their deletion, otherwise, python will raise an exception, as there is no safe deletion.

Also, `fPIC` once deleted when running Windows, will raise an exception as soon as tries to delete it again when building as shared.

So far, it's work, but is very ugly and it's a detail which is known by more experts than newbies in CCI. 

My suggestion is adding some feature to avoid that ugly big step. Some ideas that I have in mind:

- Conan automatically discards fPIC when building shared
- Keep settings and options, but manage package ID with new methods like `self.info.pure_c` and `self.info.managed_fpic`
- Add new methods to discard fPIC and compiler in middle of configure: `configure_pure_c(self)`, `configure_shared_fpic(self)`
- Add an attribute in the recipe to configure modes: `self.package_mode = ["managed_fpic", "pure_c"]`
- Re-add old remove method to settings and options: `self.settings.remove("compiler.cppstd")`
- Customize `del` operator to execute safe-delete, not raising when there is a missing object

Related to https://github.com/conan-io/conan-center-index/pull/12637#discussion_r957557619

/cc @SpaceIm 

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Yes, I agree that some higher level utilities to handle this would be great.
Some quick comments:

> Conan automatically discards fPIC when building shared

Conan being too smart and doing things like this often lead to some users being blocked because they have slightly different tools, constraints, etc. I'd prefer it to be something a bit more explicit, so users could have control


> Keep settings and options, but manage package ID with new methods like self.info.pure_c and self.info.managed_fpic

This would be a bit late, because that happens after the build. We want the toolchains to completely discard libcxx if it is a pure C project. It needs to happen earlier


> Add an attribute in the recipe to configure modes: self.package_mode = ["managed_fpic", "pure_c"]

Yes, this is something that I could see, a class attribute like ``languages = "C"``, for example, and that could be leveraged by the recipe.


My only issue is that this sounds too much as a new feature, and in this stage, I'd prefer to stabilize 2.0 before adding new features like this in 2.X. I know the syntax can be a bit ugly for pure C recipes, but it is mostly a UX inconvenience, so maybe doesn't need to be rushed.

**Comment by uilianries** [comment]
@memsharded Fine, at least we have a confirmation about a new feature. Gracias!

**Comment by stackfun** [comment]
I'ma lso looking forward to this feature :) 

We have common options like "rtti" and "sanitizer" (None/asan/tsan) that we'd like to standardize. We were considering modifying `settings.yml` but not all packages need to specify them so we decided against it.

**Comment by memsharded** [comment]
Hi! 

Automatic management of options shared+fPIC has been already implemented in ``implements`` 2.0 feature, see docs: https://docs.conan.io/2/reference/conanfile/attributes.html#implements

It applies only to ``shared`` and ``fPIC`` because those are very well known (as the title of this request says), and we had lots of experience from hundreds of recipes in ConanCenter implementing this.

So I am closing this ticket as implemented with ``implements`` in 2.0

A different story @stackfun is things like ``rtti`` and ``sanitizer``. Those don't have such large standardized practice, so it is very challenging to provide built-in implementation for them at this moment. If you think that Conan could implement some built-in utilities for them, please file a new dedicated ticket for it. It would be necessary to detail how these things are implemented in the build systems, at least the most popular ones. The recommendation for ``sanitizer`` at least could be using ``settings.yml``. It doesn't matter that some packages don't use them, because there are ways for packages to ignore the value, from using a ``None``(``null`` in ``settings.yml``) value, or doing a first level setting, or dropping it in ``configure()`` or ``package_id()``. Thanks for the feedback!

**Comment by SpaceIm** [comment]
RTTI doesn't need an option, and I've always tried to refrain contributors to add such option in conancenter recipes (even when upstream had such option in their CMakeLists). It's just one compiler option among hundreds.

I think the way to go is:
- add `tools.build:cxxflags` to `tools.info.package_id:confs` in your `global.conf`
- add `-fno-rtti` to `tools.build:cxxflags` when you want to build without rtti
